alf/environments/alf_gym3_wrapper.py

```diff
@@ -0,0 +1,407 @@
+        3. Trim ignored keys from the env info
+
+        """
+        observation = torch.as_tensor(observation)
+        if self._image_channel_first:
+            observation = nest.map_structure(
+                lambda x: x.permute(0, 3, 1, 2).contiguous(), observation)
+
+        trimmed_info = [
+            nest.prune_nest_like(info, self.env_info_spec())
+            for info in self._gym3_env.get_info()
+        ]
+
+        return ds.TimeStep(
+            step_type=torch.as_tensor(step_type),
+            reward=torch.as_tensor(reward),
+            discount=self._discount,
+            observation=observation,
+            env_id=torch.arange(self.batch_size),
+            prev_action=torch.as_tensor(action),
+            env_info=nest.map_structure(
+                lambda *values: torch.as_tensor(values), *trimmed_info))
+
+    # Implement abstract _reset
+    def _reset(self) -> ds.TimeStep:
+        """Implement ``_reset()`` for compatibility
+
+        Note that by default Gym3 environmenst cannot be reset. However, if
+        ``support_force_reset`` is set to True, action -1 will be sent to all
+        the sub-environments to reset them if the underlying environment follows
+        the convention.
+
+        Otherwise the reset will be ignored.
+
+        """
+        if self._support_force_reset:
+            self._gym3_env.act(np.array([-1] * self.batch_size))
+            self._done = [False] * self.batch_size
+        else:
+            logging.warning('reset() ignored by AlfGym3Wrapper')
+
+        reward, observation, _ = self._gym3_env.observe()
+
+        time_step = self._create_time_step(
+            reward,
+            observation,
+            step_type=[ds.StepType.FIRST] * self.batch_size,
+            # Faking actions
+            action=nest.map_structure(
+                lambda spec: spec.numpy_zeros(outer_dims=(self.batch_size, )),
+                self.action_spec()))
+
+        return time_step
+
+    # Implement abstract _step
+    def _step(self, action) -> ds.TimeStep:
+        """Implement ``_step()`` for compatibility
+
+        A very important note here is that special treatment is done at the end
+        of each episode. Unlike Gym environments, Gym3 environments do not
+        return ``done=True`` immediately. Instead, ``first=True`` is returned
+        upon the ``observe()`` of the NEXT ``act()``, which is actually the
+        FIRST FRAME of a new episode in Gym3 sense.
+
+        With this wrapper we CHANGED that definition, so that the frame with
+        ``first=True`` actually becomes the LAST FRAME of the previous episode.
+        Because the observation of the new episode can be DRAMATICALLY DIFFERENT
+        from the actual last frame of the previous episode, the previous
+        observation is returned in this case.
+
+        Therefore to summarize, effectively we will
+
+        1. Repeat the end-of-episode observation twice for each episode.
```

Comment:
Since openai ppg is directly trained on gym3 interface, which doesn't have the repeated frame at the and the skipped frame at the beginning, our training may behave slightly differently.

Reply:
Acknowledged. Right as we discussed about it, this seems the best short-term handling.

The problem with the `seed` and internal `done` approach is that after reading their code and experiment again, both of them are also "late signals", i.e. come at the next frame - the first frame of a new episode.